compiler-core/src/type_/expression.rs

```diff
@@ -1315,6 +1349,17 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
         let mut all_patterns_are_discards = true;
         let mut all_clauses_panic = true;
         for clause in clauses {
+            for pattern in clause.pattern.iter() {
+                let errors = self.check_valid_pattern_names(pattern);
+                self.errors.extend(errors);
+            }
+            for multi_pattern in clause.alternative_patterns.iter() {
+                for pattern in multi_pattern {
+                    let errors = self.check_valid_pattern_names(pattern);
+                    self.errors.extend(errors);
+                }
+            }
```

Comment:
We should not traverse the patterns an extra time for this, rather it should be done during the existing pattern analysis.